src/Lean/Elab/Tactic/LibrarySearch.lean

```diff
@@ -24,28 +24,90 @@ Implementation of the `exact?` tactic.
-          solveByElim required (exfalso := exfalso) (maxDepth := 6)
+      solveByElim required (exfalso := exfalso) (maxDepth := 6)
     let allowFailure := fun g => do
       let g ← g.withContext (instantiateMVars (.mvar g))
       return required.all fun e => e.occurs g
     match ← librarySearch goal tactic allowFailure with
     -- Found goal that closed problem
     | none =>
-      addExactSuggestion ref (← instantiateMVars (mkMVar mvar)).headBeta
+      addSuggestionIfValid ref mvar initialState
     -- Found suggestions
     | some suggestions =>
-      if requireClose then throwError
-        "`exact?` could not close the goal. Try `apply?` to see partial suggestions."
+      if requireClose then
+        let hint := if suggestions.isEmpty then "" else " Try `apply?` to see partial suggestions."
+        throwError "`exact?` could not close the goal.{hint}"
       reportOutOfHeartbeats `apply? ref
       for (_, suggestionMCtx) in suggestions do
         withMCtx suggestionMCtx do
-          addExactSuggestion ref (← instantiateMVars (mkMVar mvar)).headBeta (addSubgoalsMsg := true)
+          addSuggestionIfValid ref mvar initialState (addSubgoalsMsg := true) (errorOnInvalid := false)
       if suggestions.isEmpty then logError "apply? didn't find any relevant lemmas"
       admitGoal goal
+where
+  /-- Returns all free variables in `e` that are shadowed or not user-enterable. -/
+  collectUnprintableFVars (e : Expr) : MetaM (List FVarId) := do
+    let lctx ← getLCtx
+    let { fvarSet, .. } := Lean.collectFVars {} e
+    let unprintableFVars := fvarSet.toList.filter fun fvarId =>
+      if let some name := lctx.getRoundtrippingUserName? fvarId then
+        name.isInaccessibleUserName || name.isAnonymous || name.hasNum
+      else
+        true
+    return unprintableFVars
+
+  /--
+  Executes `tac` in `savedState` (then restores the current state). Used to ensure that a suggested
+  tactic is valid.
+
+  Remark: we don't merely elaborate the proof term's syntax because it may successfully round-trip
+  (d)elaboration but still produce an invalid tactic (see the example in #5407).
+  -/
+  evalTacticInState (savedState : Tactic.SavedState) (tac : TSyntax `tactic) : TacticM Unit := do
+    let currState ← saveState
+    savedState.restore
+    try
+      evalTactic tac
+    finally
+      currState.restore
+
+  /--
+  Suggests using the value of `goal` as a proof term if the corresponding tactic is valid at
+  `origGoal`, or else informs the user that a proof exists but is not syntactically valid.
+  -/
+  addSuggestionIfValid (ref : Syntax) (goal : MVarId) (initialState : Tactic.SavedState)
+                       (addSubgoalsMsg := false) (errorOnInvalid := true) : TacticM Unit := do
+    let proofExpr := (← instantiateMVars (mkMVar goal)).headBeta
+    let (suggestion, proofMVars) ← mkExactSuggestionSyntax proofExpr
+    let hasMVars := !proofMVars.isEmpty
+    let unprintableFVars ← collectUnprintableFVars proofExpr
```

Comment:
Will this complain about unprintable names even when they wouldn’t be part of the delaborated syntax? This is actually a common situation, e.g. with implicit parameters of the theorem that appear in this proof expression (implicitly).

Within `try?`, the most reliable and simple way I could think of was to elaborate the generated syntax without and with `expose_names` and use the first that works (see https://github.com/leanprover/lean4/pull/7133). Maybe do the same thing here.

Reply:
Thanks, I've adopted a similar approach here. I *think* `withoutErrToSorry <| withoutRecover <| ...` should be enough to catch all relevant downstream elaboration errors (these were previously not being caught).

Reply:
Looks plausible